feat: use lazy import for pylance dependencies

yliang412 opened the pull request:
## Description

This PR introduces `polars`-style lazy support and closes #1217. 

Current packages managed with `dependencies.py`:
- `numpy`
- `pandas`
- `polars`
- `torch`

Ideally, we could also add other third-party libraries and heavy-weight standard
libraries to the dependencies module.

changhiskhan commented:
**Before**
<img width="1240" alt="before" src="https://github.com/lancedb/lance/assets/70461588/0143d721-00f1-4d4d-b5ea-5e92e5b706d5">

**After**
<img width="1233" alt="after" src="https://github.com/lancedb/lance/assets/70461588/9e3be140-b7c2-433a-ab75-a893e1ac8f0c">

Let me know if I missed anything. The current code seems to pass all the PyTest tests.

westonpace commented on python/python/lance/dependencies.py:
```diff
@@ -0,0 +1,211 @@
+        "numpy": "np.",
+        "pandas": "pd.",
+        "polars": "pl.",
+    }
```
Does torch need to be in this list as `"torch": "torch."` or can it be excluded since it isn't really a prefix?

yliang412 replied:
Good catch. I will add the one for `torch`. The prefixes are basically for displaying missing imports. For instance, when I test locally, it would say the following if the `pd` prefix is registered:

<pre>ModuleNotFoundError: <b>pd</b>.DataFrame requires 'pandas' module to be installed
</pre>

versus the following if the `pd` prefix is registered:

<pre>ModuleNotFoundError: DataFrame requires 'pandas' module to be installed`
</pre>

westonpace commented on python/python/lance/dependencies.py:
```diff
@@ -0,0 +1,211 @@
+    )
+
+
+if TYPE_CHECKING:
```
What is `TYPE_CHECKING`?

yliang412 replied:
`TYPE_CHECKING` is a special constant provided by the `typing` library to third-party static type checkers. `TYPE_CHECKING` will be false at runtime.

We are also using this construct in [dataset.py](https://github.com/lancedb/lance/blob/29575e10ee5296bc2963c8e05044567b00421cd1/python/python/lance/dataset.py#L72C12-L72C12).

source: https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING

westonpace commented on python/python/lance/dependencies.py:
```diff
@@ -0,0 +1,211 @@
+
+def _check_for_numpy(obj: Any, *, check_type: bool = True) -> bool:
+    return _NUMPY_AVAILABLE and _might_be(
+        cast(Hashable, type(obj) if check_type else obj), "numpy"
```
What is this doing?

yliang412 replied:
All the `_check_for_xxx` functions are using `_might_be` function to check whether a given class is associated with a module. These functions are typically used to do a lightweight check before doing a real `isinstance` check. 

It is beyond my knowledge but the `_might_be` function is checking if "module." is in the object's method resolution order list. I was primarily following polars implementation for this part.

westonpace commented on python/python/lance/__init__.py:
```diff
@@ -37,6 +30,9 @@
 from .schema import json_to_schema, schema_to_json
 from .util import sanitize_ts
 
+ts_types = Union[datetime, pd.Timestamp, str]
```
Will `pd.Timestamp` error here if pandas is not available?

yliang412 replied:
Yes they do. I propose moving all the type definitions into `if TYPE_CHECKING` blocks.

westonpace commented on python/python/lance/arrow.py:
```diff
@@ -589,13 +587,15 @@ def _from_sequence(
                 return PandasBFloat16Array(self.data[item])
-            elif isinstance(item, np.ndarray) and item.dtype == bool:
+            elif (
+                _check_for_numpy(item)
```
Why do you need to pass in an object when calling a `_check_for_xyz` function?

yliang412 replied:
The `_check_for_xxx` is checking both if the library exist and also if the object potentially is a type from this library (following the `_might_be` discussion above).

westonpace commented on python/python/lance/dependencies.py:
```diff
@@ -0,0 +1,211 @@
+# Originally from https://github.com/pola-rs/polars/blob/ea4389c31b0e87ddf20a85e4c3797b285966edb6/py-polars/polars/dependencies.py
```
At what point are modifications substantial enough for a vendored file to receive its own copyright header I wonder?

@changhiskhan

changhiskhan replied:
i'm not sure. is there a general convention that popular oss projects follow?

yliang412 replied:
I was mainly following @wjones127's suggestion [here](https://github.com/lancedb/lance/issues/1217#issuecomment-1785487288).

wjones127 replied:
I've updated this section to make it clear the original code was taken from Polars and covered by their license, and subsequent modifications have been made by us which are covered by our license.

I've also added a section to our main license file, similar to how Apache Arrow does it: https://github.com/apache/arrow/blob/main/LICENSE.txt

westonpace reviewed (approved):
Thanks for doing this.  This looks pretty straightforward but I just have a few questions about how it is used.

yliang412 commented:
Just updated the PR according to the comments and rebased on the latest main branch.

I didn't figure out a clean way to rewrite imports that use submodules, e.g. a good way to refactor the following
```py3
from pandas.api.extensions import (
        ExtensionArray,
        ExtensionDtype,
        register_extension_dtype,
    )
```

changhiskhan commented:
@yliang412 are you able to resolve the conflicts and we can merge this in? thanks

yliang412 commented:
@changhiskhan Yep, I just resolved all the conflicts. It might be a good idea to let other contributors know that the lazy dependency management module is available.

wjones127 commented:
Oddly, I don't see much improvement in import time. I think this is because we subclass `pyarrow.dataset.Dataset`, and that import brings in numpy and Pandas.

## Before

![Screenshot 2024-01-11 at 12 54 32 PM](https://github.com/lancedb/lance/assets/5488879/60eaec49-c582-4649-9421-014468ad8f62)


## After

![Screenshot 2024-01-11 at 12 54 18 PM](https://github.com/lancedb/lance/assets/5488879/fcfb7690-e052-470d-b05a-183fd455c6d1)

As a follow up, we should explore making `LanceDataset` not be a subclass of `pyarrow.dataset.Dataset`.

wjones127 reviewed (approved):
This looks good, and I am indeed seeing an improvement on import times. The one change we needed was to add a section to our license.

There are a few follow-ups I could see us doing for this:

1. Add a benchmark for import time, so we can detect regressions.
2. Make `LanceDataset` not inherit from `pyarrow.dataset.Dataset`, so we don't have to import that. This seems to be the biggest chunk of our import time now.
